### PR TITLE
Long-Term Occupancy Forecast (ML)

### PR DESCRIPTION
**Pull request by DottL**
# Long-Term Forecast (ML)

## Summary

Adds the long-term occupancy forecasting ML pipeline: two-stage hybrid model, training/evaluation/promotion/inference scripts, long-term-specific baselines, coverage-gated baseline comparison, and a horizon-stratified promotion gate.

---

## New model — two-stage XGBoost hybrid (Long-Term)

- **Stage 1:** 4-week rolling baseline per `(lot_id, academic_period, day_of_week, hour)`. Sparse groups fall back to global `(day_of_week, hour)` mean.
- **Stage 2:** XGBoost learns the deviation from baseline using `historical_baseline`, `days_ahead`, `week_of_semester`, `is_campus_open`, sin/cos hour/day encodings, and categorical `lot_id`/`semester`/`academic_period`. Target is `deviation = actual_rate - historical_baseline`.
- Three quantile models (median, 10th, 90th percentile) for confidence intervals. Cold-start lots get widened CIs via `COLD_START_CI_MULTIPLIER`.

## New Long-Term Feature Engineering

- `compute_baseline()` — rolling window aggregation with cold-start fallback
- `prepare_training_features()` — simulates `days_ahead` in [1, 7] per row; included as a feature for when lag features are added, but currently has minimal effect since all other features are calendar-only
- `prepare_inference_features()` — generates `(lot × date × hour)` grid for production inference; calendar features derived directly from `academic_calendar.py` per target date

## New pipeline scripts

| Script | Purpose |
|--------|---------|
| `scripts/train_long_term.py` | Loads data, computes baseline, trains model, logs to `sharkpark-long-term` MLflow experiment |
| `scripts/evaluate_long_term.py` | Horizon-stratified MAE table; three-gate promotion (baselines → horizon targets → production improvement) |
| `scripts/promote_long_term.py` | Registers as `long-term-production` with `@production` alias |
| `scripts/predict_long_term.py` | Batch inference writing to `predictions_long_term`; supports `--days-ahead N`, `--write-local`, `--data-path` |
| `scripts/check_long_term_predictions.py` | DB inspection with `--lot`, `--date`, `--days-ahead`, `--summary` filters |

## Long-term baseline classes (`src/models/baselines.py`)

- **`HistoricalAverageBaseline`** — predicts the Stage 1 baseline rate (equivalent to "deviation = 0"). A candidate that fails to beat this is not learning useful adjustments.
- **`SameDayLastWeekBaseline`** — carries the most recent same-slot `(lot_id, day_of_week, hour)` observation forward; slots with no prior observation fall back to the global median of actual rates.
- **`MajorityClassBaseline.evaluate`** — extended with an optional `actuals` parameter; long-term has no `target_occupancy_rate` column so actual rates are passed in directly instead of read from the test features.

## Promotion gates

Long-term must pass **all three**, in order:
1. **Baseline gate** — beat all active baselines on MAE (coverage-gated, see below).
2. **Horizon gate** — meet per-horizon MAE targets: flat < 0.15 across all days. The current long-term model is calendar-only (no same-day lag features), so it has no informational advantage at Day 1 vs Day 7 — tiered targets only make sense once a `weekly_trend` feature is added (planned, requires more real data). (see Model_Design.md)
3. **Production gate** — ≥5% MAE improvement OR ≥3pp directional accuracy gain vs current production.

## Coverage-gated long-term baselines (`src/evaluation/compare.py`)

`compare_against_long_term_baselines()` evaluates each baseline only when there is enough data to make it meaningful. Each baseline has a different minimum data requirement:

| Baseline | Included when | Minimum obs per slot |
|----------|--------------|----------------------|
| `MajorityClass` | ≥30% of slots covered | ≥2 |
| `SameDayLastWeek` | ≥50% of slots covered | ≥1 |
| `HistoricalAverage` | ≥60% of slots covered | ≥4 |

If overall coverage falls below 30%, baseline comparison is skipped entirely.

## Feature changes

- Dropped `sin_hour`/`cos_hour`/`sin_day`/`cos_day` from short-term features — redundant with raw `hour`/`day_of_week` for tree models that only predict 7am–9pm (no wrap-around).
- Dropped raw `hour`/`day_of_week` from long-term model features — redundant with `sin_hour`/`cos_hour`/`sin_day`/`cos_day`, which are load-bearing for long-term since there are no lag features. (Columns still carried through feature prep for baseline joins, just not fed to XGBoost.)

## Modified shared files

| File | Change |
|------|--------|
| `src/models/base.py` | New — `BaseXGBoostModel` shared by both models |
| `src/features/base.py` | Refactored from short-term — shared base feature utilities used by both pipelines; adds `merge_real_synthetic()` |
| `src/evaluation/compare.py` | Adds horizon MAE targets, long-term baseline comparison with coverage gating, and shared promotion helpers |
| `src/utils/mlflow_utils.py` | New — shared MLflow helpers extracted from evaluate scripts |
| `src/config.py` | New long-term constants for model name, horizon, baseline window, and cold-start CI multiplier |
| `src/data/db.py` | Adds `load_historical_snapshots()` and `write_long_term_predictions()`; renames `write_predictions` → `write_short_term_predictions` |
| `src/models/short_term.py` | Refactored to extend `BaseXGBoostModel` |
| `src/features/short_term.py` | Refactored to use `prepare_base_features()`; switched time encoding from sin/cos to raw `hour`/`day_of_week` |
| `scripts/predict_short_term.py` | Added missing-lot warning to `_build_prediction_df()` |
| Short-term scripts + tests | Renamed to `*_short_term.py` to make room for long-term equivalents |
| `README.md` | Updated with actual short-term/long-term commands and `uv` install instructions |
| `Model_Design.md` | Updated feature list to reflect sin/cos-only time encoding for long-term |

## Tests (+50, total 247)

- **`tests/features/test_long_term.py`** (new, 12 tests) — `compute_baseline` confidence filtering and coverage fallback; `prepare_training_features` output columns and deviation correctness
- **`tests/models/test_long_term.py`** (new, 12 tests) — train/save/load/MLflow round-trip, untrained-model error paths, quantile ordering, cold-start CI widening; LOW-confidence rows not filtered at train time (downweighted via sample weights instead)
- **`tests/models/test_baselines.py`** — adds 8 tests for `HistoricalAverageBaseline` + `SameDayLastWeekBaseline` + `MajorityClassBaseline.actuals` kwarg
- **`tests/evaluation/test_compare.py`** — adds 14 tests net: 10 for `meets_horizon_targets`, `build_horizon_promotion_reason` (horizon gate), and `compare_against_long_term_baselines`; plus new `TestBuildPromotionReason` class (5 tests) with `TestCompareModels` restructured (−1 net)
- **`tests/scripts/test_workflow_long_term.py`** (new, 4 tests) — full train → evaluate → promote → predict integration, production-gate regression check, missing-data and MLflow error paths

---

### CODE REVIEW

File: services/ml/src/features/long_term.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+__all__ = [
+    "compute_baseline",
+    "prepare_training_features",
+    "prepare_inference_features",
+]
+
+# Minimum unique days per (lot_id, academic_period, day_of_week, hour) group
+# before trusting its mean; sparse groups fall back to global (day_of_week, hour)
+# Must be ≤2 to work within short academic periods (dead_week, midterms)
+_MIN_COVERAGE_DAYS = 2
```
Comment (chuckmilton) [inline services/ml/src/features/long_term.py:49]: If a lot has only LOW-confidence snapshots, the confidence filter in `compute-baseline()` drops all its data and this group falls through to the global `(day_of_week, hour)` fallback silently. Since this is intentional, I think it would be nice to add a brief comment here so others don't think it's a bug in the future.

File: services/ml/src/models/long_term.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                train_features, synthetic_weight, cold_start_weight
+            )
+        else:
+            sample_weight = None
+
+        # Encode categoricals
+        self._fit_category_mappings(train_features)
+        X_train, y_train = self._prepare_xy(train_features)
+
+        # TODO: tune hyperparams with Optuna or similar once we have enough real data
```
Comment (chuckmilton) [inline services/ml/src/models/long_term.py:162]: The `TODO` comment is fine to leave, but to make sure it doesn't get buried, we should open a tracking issue. Once we have enough real data, this will be a meaningful lever.

**Review by chuckmilton — APPROVED** [review]
Amazing job on the Long-Term Forecast feature! The two-stage hybrid architecture (rolling baseline + XGBoost deviation) is a smart design. I agree that predicting residuals is easier to fit and generalizes better than raw occupancy. The `BaseXGBoostModel` extraction eliminates real duplication without over-abstracting. Coverage-gated promotion with 3 independent gates (baselines, horizon targets, production improvement) is very thorough and prevents weak models from shipping.

The following here are also implemented very well:
- Baseline computation with confidence filtering + sparse-group global fallback is well-calibrated for short academic periods (dead week, midterms)
- `days_ahead` simulation during training lets the model learn accuracy degradation across the 7-day horizon
- Cold-start CI widening via spread-based scaling (not boundary shifting) is correct
- Dropping sin/cos for raw `hour`/`day_of_week` in short-term is the right call for tree models
- New tests with good coverage of edge cases, round-trips, and full workflows

Btw, this is what I got from the test results (unrelated to PR): 243 passed, 4 failed. The 4 failures are pre-existing in `TestFetchLots` (`test_synthetic.py`). Neither the test file nor `synthetic.py` were modified in this PR. The mock doesn't intercept the DATABASE_URL env-var check that runs before the patched code path. Worth a separate fix.

Overall, LGTM 💯
